### ISSUE TITLE
Some cublas functions don't handle inputs with zero strides

### ISSUE THREAD

**Issue by shchur** [pytorch/pytorch#29984]
## To Reproduce

```python
import torch
import torch.nn as nn

torch.set_default_tensor_type('torch.cuda.FloatTensor')
x = nn.Parameter(torch.ones(2, 2))
(x @ torch.ones(2)).sum().backward()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-11-c3b66f275e9a> in <module>()
      1 x = nn.Parameter(torch.ones(2, 2))
----> 2 (x @ torch.ones(2)).sum().backward()

1 frames
/usr/local/lib/python3.6/dist-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
     97     Variable._execution_engine.run_backward(
     98         tensors, grad_tensors, retain_graph, create_graph,
---> 99         allow_unreachable=True)  # allow_unreachable flag
    100 
    101 

RuntimeError: cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:120
```
## Expected behavior

No exception is raised.

## Environment
```
Collecting environment information...
PyTorch version: 1.3.1+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1+cu100
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.4.2+cu100
[conda] Could not collect
```

This code was run on Google Colab. The bug also reproduces on my server (Ubuntu 16.04LTS, 4x GTX1080Ti GPUs).

## Additional context

The bug doesn't occur if:
1. If I replace `sum()` with `mean()`
```python
(x @ torch.ones(2)).mean().backward()
```

2. If I use a different shape of the tensor
```python
(x @ torch.ones(2, 1)).sum().backward()
```

3. If I run the code on CPU

Therefore, I believe that this is not intended behavior.

cc @ezyang @gchanan @zou3519 @jerryzh168 @SsnL @albanD @gqchen

**Comment by ngimel** [comment]
Simpler repro
```
import torch

x = torch.ones(2, device="cuda")
b = torch.randn(1, device="cuda").expand(2)
print(x.ger(b)) # cublas runtime error : an invalid numeric value was used as an argument at /pytorch/aten/src/THC/THCBlas.cu:120
x = torch.ones(2,2, device="cuda")
print( x @ b) # at::cuda::blas::gemv<float> argument incx must be positive and less than 2147483647 but got 0
```
some cublas functions (gemv and ger at least) don't handle broadcasted arguments, and thus broadcasted arguments should be expanded before calling into them. Note, above repro does not produce errors on cpu. 
cc: @ptrblck @csarofeen

**Comment by mruberry** [comment]
@ezyang would like a better error message

**Comment by zasdfgbnm** [comment]
@mruberry @ngimel I am self assigning this.

**Comment by zasdfgbnm** [comment]
The bug is actually inside legacy code like `at::native::legacy::cuda::_th_mv`. I won't fix the legacy code. But I will migrate it to aten native first then fix. I will self assign related migration issues.

cc: @VitalyFedyunin (Don't forget the gather migration :) )
cc: @csarofeen @ptrblck

**Comment by VitalyFedyunin** [comment]
`at::native::legacy::cuda::_th_mv` moved to aten in #30898

**Comment by Balandat** [comment]
Is there a timeline for the `gemv` fix? Seems odd that we've known for more than half a year that  matrix-vector multiplication is broken on the GPU.

**Comment by zasdfgbnm** [comment]
@kshitij12345 Is this fixed by #38321?

**Comment by ngimel** [comment]
`gemv` is fixed, `ger` is not (fun fact - in many cases it is faster to do `ger` by broadcasted multiply which automatically handles zero strides, than by calling into cublas ger).
